### PR TITLE
[FEATURE] New gas price mechanism - Minimum gas price

### PR DESCRIPTION
**Pull request by Bagaric**
Adjusts `timenode-core` to the latest changes in the EAC protocol https://github.com/ethereum-alarm-clock/ethereum-alarm-clock/pull/149.

If the network conditions during execution require a gas price higher than set in the scheduled transaction, the **TimeNodes now use the gas price optimal for those network conditions.**

**Additional:**
- Adds an additional economic strategy parameter _maxGasSubsidy_ which defines TimeNode's willingness to subsidize the gas price in case of spikes.

**maxGasSubsidy** - number 0-100
- Percentage of the original transaction gas price which the TimeNode is willing to add to the gasPrice in case of network gas price spikes.

If no _maxGasSubsidy_ is set in the economic strategy, the TimeNode will use current best network gas price in case the current best network gas price is higher than the gas price set in the transaction.

_Note: TimeNode still guarantees that it will send at least the minimum amount of gas set in the transaction._

**UPDATE:**
This PR also adds a check which stops the TimeNodes from executing transactions which bring a negative profit should the gas cost go up significantly at the time of execution.

### CODE REVIEW

File: src/EconomicStrategy/EconomicStrategyHelpers.ts [hunk 1]
```diff
@@ -1,13 +1,17 @@
 import { IEconomicStrategy } from './IEconomicStrategy';
 import Config from '../Config';
+import { BigNumber } from '../../node_modules/bignumber.js';
```
Comment (lsaether) [inline src/EconomicStrategy/EconomicStrategyHelpers.ts]: Can we shorten the import to just 'bignumber.js'?

File: src/EconomicStrategy/EconomicStrategyHelpers.ts [hunk 2]
```diff
@@ -22,7 +26,7 @@ const exceedsMaxDeposit = (txRequest: any, economicStrategy: IEconomicStrategy)
  * Checks if the balance of the TimeNode is above a set limit.
  * @param {Config} config TimeNode configuration object.
  */
-const isAboveMinBalanceLimit = async (config: Config) => {
+const isAboveMinBalanceLimit = async (config: Config): Promise<boolean | BigNumber> => {
```
Comment (lsaether) [inline src/EconomicStrategy/EconomicStrategyHelpers.ts]: These should only return boolean type

File: src/EconomicStrategy/EconomicStrategyHelpers.ts [hunk 3]
```diff
@@ -38,7 +42,10 @@ const isAboveMinBalanceLimit = async (config: Config) => {
  * @param {TransactionRequest} txRequest Transaction Request object to check.
  * @param {IEconomicStrategy} economicStrategy Economic strategy configuration object.
  */
-const isProfitable = async (txRequest: any, economicStrategy: IEconomicStrategy) => {
+const isProfitable = async (
+  txRequest: any,
+  economicStrategy: IEconomicStrategy
+): Promise<boolean | BigNumber> => {
```
Comment (lsaether) [inline src/EconomicStrategy/EconomicStrategyHelpers.ts]: only return Promise<boolean> type

File: src/EconomicStrategy/EconomicStrategyHelpers.ts (hunk 1 above)
Comment (Bagaric) [reply src/EconomicStrategy/EconomicStrategyHelpers.ts]: Yeah, this was auto generated by VScode

**Comment by kosecki123** [comment]
I think this the part of the story and not sure if static percentage will cover all cases, especially this case.

1) TimeNode claimed with deposit = 0.1 ETH
2) At the time of execution, the gasPrice went to 100gwei from the minimal set by the scheduler as 20gwei
3) Regardless the settings (because of deposit at risk) TimeNode need to execute in order to not lose the deposit. The worst case is that the `cost = deposit + reward + reimbursement` so TimeNode loses deposit but tx has been executed.

Example 1) 

Gas amount = 100k
Reward = 100gwei
Deposit = 500gwei

MinGasPrice = 1gwei
CurrentGasPrice = 50gwei

Reimbursement = 1gwei * 100k = 100k gwei
CurrentGasCost = 50gwei * 100k = 5000k gwei

In this case, because the CurrentGasCost > Deposit + Reward + Reimbursement the TimeNode should wait or eventually `not execute`

Example 2)

Gas amount = 100k
Reward = 100gwei
Deposit = 50000 gwei

MinGasPrice = 1gwei
CurrentGasPrice = 50gwei

Reimbursement = 1gwei * 100k = 100k gwei
CurrentGasCost = 50gwei * 100k = 5000k gwei

In this case, because the CurrentGasCost < Deposit + Reward + Reimbursement the TimeNode should wait or eventually `execute`

File: src/EconomicStrategy/EconomicStrategyHelpers.ts [hunk 4]
```diff
@@ -114,4 +114,30 @@ const getExecutionGasPrice = async (txRequest: any, config: Config): Promise<Big
...
+
+  const gasPrice = await this.getExecutionGasPrice(txRequest, config);
+  const gasAmount = config.util.calculateGasAmount(txRequest);
+  const reimbursement = txRequest.gasPrice.times(gasAmount);
+  const deposit = isClaimedByMe ? txRequest.requiredDeposit : new BigNumber(0);
+
+  const paymentModifier = await txRequest.claimPaymentModifier();
+  const reward = txRequest.bounty.times(paymentModifier);
+
+  const gas = gasPrice.times(gasAmount);
```
Comment (lsaether) [inline src/EconomicStrategy/EconomicStrategyHelpers.ts]: Rename gas => gasCost for clarity

File: src/EconomicStrategy/EconomicStrategyHelpers.ts [hunk 5]
```diff
@@ -71,4 +80,64 @@ const shouldClaimTx = async (txRequest: any, config: Config) => {
...
+  const gasAmount = config.util.calculateGasAmount(txRequest);
+  const reimbursement = txRequest.gasPrice.times(gasAmount);
+  const deposit = isClaimedByMe ? txRequest.requiredDeposit : new BigNumber(0);
+
+  const paymentModifier = await txRequest.claimPaymentModifier();
+  const reward = txRequest.bounty.times(paymentModifier);
+
+  const gasCost = gasPrice.times(gasAmount);
+
+  if (gasCost.greaterThan(deposit.plus(reward).plus(reimbursement))) {
```
Comment (kosecki123) [inline src/EconomicStrategy/EconomicStrategyHelpers.ts]: Please move this `deposit.plus(reward).plus(reimbursement)` to separate variable. Also add `debug` logging with gasCost and 'expectedReward' (or whatever name for above)

Comment (kosecki123) [reply src/EconomicStrategy/EconomicStrategyHelpers.ts]: Added this in `Added logging to shouldExecuteTx` commit

**Pull request merged**